### PR TITLE
Averaging Entities' Properties

### PR DESCRIPTION
**Pull request by omermadmon**
Implemented methods for averaging entities' properties:
* A method for calculating the average of list of entities over a single property - this returns the average value.
* A method for calculating the averages of list of entities over a **set** of properties - this return a map where the key indicates the property and the value is the average value of this property.

These method will be used for providing statistical data over multiple simulation.

### CODE REVIEW

File: src/main/java/com/google/research/bleth/utils/Queries.java [hunk 1]
```diff
@@ -82,4 +85,66 @@
...
         return result;
     }
+
+    /**
+     * Return the average of entities property value. Does not include entities which don't have property.
+     * @param entities is a list of entities.
+     * @param property is the property to average.
+     * @return The average value (or Nan if entities is empty or no entity with value exists).
+     */
+    public static double Average(List<Entity> entities, String property) {
```
Comment (OrpazFishel) [inline src/main/java/com/google/research/bleth/utils/Queries.java:95]: Can you use this method to implement the second method (or vice versa)?

File: src/test/java/com/google/research/bleth/utils/QueriesIT.java [hunk 2]
```diff
@@ -352,8 +369,154 @@ public void provideNonExistingForeignKey_shouldRetrieveEmptyList() {
...
+        List<Entity> entities = retrieveEntities(ENTITY_KIND);
+        Set<String> properties = new HashSet<>(Arrays.asList(PROPERTY_A, PROPERTY_B, PROPERTY_C));
+        Map<String, Double> expectedResult = new HashMap<>();
+        expectedResult.put(PROPERTY_A, 1.0);
+        expectedResult.put(PROPERTY_B, 1.5);
+        expectedResult.put(PROPERTY_C, 2.0);
+
+        Map<String, Double> actualResult = Queries.Average(entities, properties);
+
+        for (String property : properties) {
```
Comment (OrpazFishel) [inline src/test/java/com/google/research/bleth/utils/QueriesIT.java:453]: Consider using `assertThat(actualResult).containsExactlyEntriesIn(expectedResult);`.

File: src/main/java/com/google/research/bleth/utils/Queries.java [hunk 3]
```diff
@@ -82,4 +85,66 @@
...
+                .mapToDouble(entity -> (double) entity.getProperty(property))
+                .average()
+                .orElse(Double.NaN);
+    }
+
+    /**
+     * Return the average of entities multiple properties values (an average for each property).
+     * For each property, include only entities which have the property.
+     * @param entities is a list of entities.
+     * @param properties is a list of properties.
```
[the remark this answers is not in the document]
Comment (omermadmon) [reply src/main/java/com/google/research/bleth/utils/Queries.java:110]: Will do, and I also think of adding a test case where the property cannot be casted to double.

File: src/test/java/com/google/research/bleth/utils/QueriesIT.java (hunk 2 above)
Comment (omermadmon) [reply src/test/java/com/google/research/bleth/utils/QueriesIT.java:453]: Nice!

File: src/main/java/com/google/research/bleth/utils/Queries.java [hunk 4]
```diff
@@ -82,4 +87,33 @@
...
     }
+
+    /**
+     * Return the average of entities property value. Does not include entities which don't have property.
+     * @param entities is a list of entities.
+     * @param property is the property to average.
+     * @return The average value (or Nan if entities is empty or no entity with value exists).
+     * @throws ClassCastException if property value cannot be casted to double (for some entity).
+     */
+    public static double Average(List<Entity> entities, String property) throws ClassCastException {
```
Comment (nolaviz) [inline src/main/java/com/google/research/bleth/utils/Queries.java:98]: Method names should use lowerCamelCase.

File: src/main/java/com/google/research/bleth/utils/Queries.java [hunk 5]
```diff
@@ -82,4 +87,33 @@
...
+    /**
+     * Return the average of entities multiple properties values (an average for each property).
+     * For each property, include only entities which have the property.
+     * @param entities is a list of entities.
+     * @param properties is a list of properties.
+     * @return a map where keys indicating properties' names and values indicating the averages (for each entry, Nan
+     * value indicating no entity with property exists within entities list).
+     * @throws ClassCastException if one of properties values cannot be casted to double (for some entity).
+     */
+    public static Map<String, Double> Average(List<Entity> entities, Set<String> properties) throws ClassCastException {
```
Comment (nolaviz) [inline src/main/java/com/google/research/bleth/utils/Queries.java:115]: Ditto (lowerCamelCase).

File: src/test/java/com/google/research/bleth/utils/QueriesIT.java [hunk 6]
```diff
@@ -352,8 +370,165 @@ public void provideNonExistingForeignKey_shouldRetrieveEmptyList() {
...
+        expectedResult.put(PROPERTY_A, 1.0);
+        expectedResult.put(PROPERTY_B, 1.5);
+        expectedResult.put(PROPERTY_C, Double.NaN);
+
+        Map<String, Double> actualResult = Queries.Average(entities, properties);
+
+        assertThat(actualResult).containsExactlyEntriesIn(expectedResult);
+    }
+
     @After
```
Comment (nolaviz) [inline src/test/java/com/google/research/bleth/utils/QueriesIT.java:494]: Please be consistent: either leave this at the end of the class, or move it to the beginning.
Don't leave it in the middle.

**Review by nolaviz — APPROVED** [review]

**Pull request merged**